libcaf_core/caf/intrusive/linked_list.hpp

```diff
@@ -0,0 +1,285 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/config.hpp"
+#include "caf/intrusive/forward_iterator.hpp"
+
+#include <memory>
+#include <utility>
+
+namespace caf::intrusive {
+
+/// A singly-linked list implementation.
+template <class T>
+class linked_list {
+public:
+  // -- member types ----------------------------------------------------------
+
+  using value_type = T;
+
+  using node_type = typename value_type::node_type;
+
+  using node_pointer = node_type*;
+
+  using pointer = value_type*;
+
+  using const_pointer = const value_type*;
+
+  using reference = value_type&;
+
+  using const_reference = const value_type&;
+
+  using unique_pointer = std::unique_ptr<T>;
+
+  using iterator = forward_iterator<value_type>;
+
+  using const_iterator = forward_iterator<const value_type>;
+
+  // -- static utility functions -----------------------------------------------
+
+  /// Casts a node type to its value type.
+  static pointer promote(node_pointer ptr) noexcept {
+    return static_cast<pointer>(ptr);
+  }
+
+  // -- constructors, destructors, and assignment operators -------------------
+
+  linked_list() noexcept {
+    init();
+  }
+
+  linked_list(linked_list&& other) noexcept {
+    if (other.empty()) {
+      init();
+    } else {
+      head_.next = other.head_.next;
+      tail_.next = other.tail_.next;
+      tail_.next->next = &tail_;
+      size_ = other.size_;
+      other.init();
+    }
+  }
+
+  linked_list& operator=(linked_list&& other) noexcept {
+    clear();
+    if (other.empty()) {
+      init();
+    } else {
+      head_.next = other.head_.next;
+      tail_.next = other.tail_.next;
+      tail_.next->next = &tail_;
+      size_ = other.size_;
+      other.init();
+    }
+    return *this;
+  }
+
+  linked_list(const linked_list&) = delete;
+
+  linked_list& operator=(const linked_list&) = delete;
+
+  ~linked_list() {
+    clear();
+  }
+
+  // -- observers -------------------------------------------------------------
+
+  /// Returns the accumulated size of all stored tasks.
+  size_t size() const noexcept {
+    return size_;
+  }
+
+  /// Returns whether the queue has no elements.
+  bool empty() const noexcept {
+    return size() == 0;
+  }
+
+  // -- modifiers -------------------------------------------------------------
+
+  /// Removes all elements from the queue.
+  void clear() noexcept {
+    typename unique_pointer::deleter_type fn;
+    drain(fn);
+  }
+
+  /// Removes the first element from the queue and returns it.
+  unique_pointer pop_front() {
+    unique_pointer result;
+    if (!empty()) {
+      auto ptr = promote(head_.next);
+      head_.next = ptr->next;
+      if (--size_ == 0) {
+        CAF_ASSERT(head_.next == &(tail_));
+        tail_.next = &(head_);
```

Comment:
Not sure about the parentheses?
```suggestion
        CAF_ASSERT(head_.next == &tail_);
        tail_.next = &head_;
```